sch_cake.c

```diff
@@ -1637,28 +1665,50 @@ static u8 cake_handle_diffserv(struct sk_buff *skb, u16 wash)
 {
 	int wlen = skb_network_offset(skb);
 	u8 dscp;
+	unsigned char buf[2];
+	struct ipv6hdr *ip6h;
+	struct iphdr *iph;
 
 	switch (cake_skb_proto(skb)) {
 	case htons(ETH_P_IP):
-		wlen += sizeof(struct iphdr);
-		if (!pskb_may_pull(skb, wlen) ||
-		    skb_try_make_writable(skb, wlen))
+		/* ToS is in the second byte of iphdr */
+		iph = skb_header_pointer(skb, wlen, sizeof(buf), &buf);
```

Comment:
This creates a 'struct iphdr' pointer that points to a data area that is too small, but the fact that it is can be easy to miss. So I'd prefer if you not do that. Instead, use a pointer of a different type to store the return value from skb_header_pointer (say, use a __be16 value for buf, and a pointer of the same type), and then use an explicit cast and a comment when passing it to ipv{4,6}_get_dsfield(). That way it's obvious that something strange is going on, and no one is likely to, say, wonder why the 'iph' pointer is not just reused in the call to ipv{4,6}_change_dsfield() later :)